subsys/prefetch/src/main/java/org/commonjava/indy/subsys/prefetch/PrefetchManager.java

```diff
@@ -104,32 +144,40 @@ public void registerPrefetchStores( @Observes final ArtifactStorePostUpdateEvent
     void triggerWorkers()
     {
         logger.trace( "Trigger works now" );
+
+        //TODO: should use a separated thread here to loop this resource working to avoid main thread holding here.
         while ( frontier.hasMore() )
         {
-            Map<RemoteRepository, List<ConcreteResource>> resources = frontier.remove( config.getBatchSize() );
+            Map<RemoteRepository, List<RescanableResourceWrapper>> resources = frontier.remove( config.getBatchSize() );
             logger.trace( "Start to trigger threads to download {}", resources );
-            executor.execute( new PrefetchWorker( transfers, frontier, resources, PrefetchManager.this, logger ) );
-        }
-    }
-
-    private List<String> buildPahts( final RemoteRepository repository )
-    {
-        for ( ContentListBuilder builder : listBuilders )
-        {
-            if ( repository.getPrefetchListingType().equals( builder.type() ) )
+            boolean hasResourceDownload = false;
+            for ( Map.Entry<RemoteRepository, List<RescanableResourceWrapper>> resource : resources.entrySet() )
+            {
+                if ( resource.getValue() != null && !resource.getValue().isEmpty() )
+                {
+                    hasResourceDownload = true;
+                    break;
+                }
+            }
+            if ( hasResourceDownload )
             {
-                logger.trace( "Use {} for {}", builder, repository.getName() );
-                return builder.buildPaths( repository );
+                prefetchExecutor.execute( new PrefetchWorker( transfers, frontier, resources, PrefetchManager.this,
+                                                              specialPathManager ) );
             }
         }
-
-        // By default, we will use html content list builder if no builder matched.
-        return new HtmlContentListBuilder().buildPaths( repository );
     }
 
+
     @PreDestroy
     public void stopPrefeching()
     {
-        frontier.stopSchedulingMore();
+        if ( config.isEnabled() )
+        {
+            stopped = true;
+            frontier.stopSchedulingMore();
```

Comment:
Do we need to await the existing, running downloads? We don't want to end up with partial downloads.

Reply:
Sure. Need to let the prefetch-work pool to wait the shutdown finished. Will add it.